addons/netfox/rollback/rollback-synchronizer.gd

```diff
@@ -263,34 +314,92 @@ func _submit_input(input: Dictionary, tick: int):
-		pass
-	
-	if tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state >= 0:
+func _submit_full_state(received_state: Dictionary, received_tick: int):
+	if received_tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state_tick >= 0:
 		# State too old!
-		_logger.error("Received state for %s, rejecting because older than %s frames" % [tick, NetworkRollback.history_limit])
+		_logger.error("Received state for %s, rejecting because older than %s frames" % [received_tick, NetworkRollback.history_limit])
 		return
 
-	var sender = multiplayer.get_remote_sender_id()
+	var sender_id = multiplayer.get_remote_sender_id()
 	var sanitized = {}
-	for property in state:
-		var property_entry = _property_cache.get_entry(property)
-		var value = state[property]
-		var state_owner = property_entry.node.get_multiplayer_authority()
+
+	for property_name in received_state:
+		var property_entry = _property_cache.get_entry(property_name)
+		var value = received_state[property_name]
+		var state_owner_id = property_entry.node.get_multiplayer_authority()
 		
-		if state_owner != sender:
+		if state_owner_id != sender_id:
 			_logger.warning("Received state for node owned by %s from %s, sender has no authority!" \
-				% [state_owner, sender])
+				% [state_owner_id, sender_id])
 			continue
 		
-		sanitized[property] = value
+		sanitized[property_name] = value
+	
+	# Duplicates the previous state(s), including the current one
+	# Also fixes packet loss (e.g. a state missing)
+	if (NetworkRollback.enable_state_diffs && _latest_state_tick != -1):
+		for picked_tick in range(_latest_state_tick, received_tick - 1):
+			_states[picked_tick + 1] = _states[picked_tick].duplicate(true)
 	
 	if sanitized.size() > 0:
-		_states[tick] = PropertySnapshot.merge(_states.get(tick, {}), sanitized)
-		# _latest_state = max(_latest_state, tick)
-		_latest_state = tick
+		_states[received_tick] = PropertySnapshot.merge(_states.get(received_tick, {}), sanitized)
+		_latest_state_tick = received_tick
+		
+		if (NetworkRollback.enable_state_diffs and not _has_received_full_state):
+			_has_received_full_state = true
+			_receive_full_state_ack.rpc_id(1, received_tick)
+	else:
+		_logger.warning("Received invalid state from %s for tick %s" % [sender_id, received_tick])
+
+## The received state ALWAYS has missing properties.
+@rpc("any_peer", "unreliable_ordered", "call_remote")
+func _submit_diff_state(received_state: Dictionary, received_tick: int):
+	if received_tick < NetworkTime.tick - NetworkRollback.history_limit and _latest_state_tick >= 0:
+		# State too old!
+		_logger.error("Received state for %s, rejecting because older than %s frames" % [received_tick, NetworkRollback.history_limit])
+		return
+
+	if (_states.is_empty() || _latest_state_tick == -1):
+		_logger.error("Received a state without any properties from %s, but missing previous states!" % multiplayer.get_remote_sender_id())
```

Comment:
I don't exactly follow this error message - what do you mean *"without any properties"*? There's no check here whether the received diff state has any properties.

Reply:
I cannot find this codeblock in the latest version, but this condition was a necessary inclusion for the scenario where the client received a diff state, without having previously received a full state. In the case of the full state being victim of packet loss (since it is `"unreliable_ordered"`)

I agree that the error message is unhelpful and vague.

I suggest re-placing this condition block, and renaming the logger error message to something like:
"Received diff state, but there is no full state received prior."

Reply:
Thinking more about this, this if condition shouldn't ever trigger because if the first full state is lost, it doesn't send an ack to start sending diff states. If you want to be safe, add the condition and we remove it in a subsequent PR if no one triggers it, otherwise it is better removed for clean code :+1: